src/shared/pam-util.c

```diff
@@ -8,23 +8,34 @@
 #include "errno-util.h"
 #include "macro.h"
 #include "pam-util.h"
+#include "string-util.h"
 
-int pam_log_oom(pam_handle_t *handle) {
-        /* This is like log_oom(), but uses PAM logging */
-        pam_syslog(handle, LOG_ERR, "Out of memory.");
-        return PAM_BUF_ERR;
-}
+int pam_syslog_errno(pam_handle_t *handle, int level, int error, const char *format, ...) {
+        va_list ap;
+
+        LOCAL_ERRNO(error);
+
+        va_start(ap, format);
+        pam_vsyslog(handle, LOG_ERR, format, ap);
+        va_end(ap);
 
-int pam_bus_log_create_error(pam_handle_t *handle, int r) {
-        /* This is like bus_log_create_error(), but uses PAM logging */
-        pam_syslog(handle, LOG_ERR, "Failed to create bus message: %s", strerror_safe(r));
-        return PAM_BUF_ERR;
+        return error == -ENOMEM ? PAM_BUF_ERR : PAM_SERVICE_ERR;
 }
 
-int pam_bus_log_parse_error(pam_handle_t *handle, int r) {
-        /* This is like bus_log_parse_error(), but uses PAM logging */
-        pam_syslog(handle, LOG_ERR, "Failed to parse bus message: %s", strerror_safe(r));
-        return PAM_BUF_ERR;
+int pam_syslog_pam_error(pam_handle_t *handle, int level, int error, const char *format) {
+        /* This wraps pam_syslog() but supports specifying %m at the very end (only!)
+         * to be replaced with pam_strerror(). */
+
+        if (endswith(format, "%m")) {
+                const char *t = strndupa_safe(format, strlen(format) - 2);
+                assert(!strchr(t, '%'));
+                pam_syslog(handle, level, "%s%s", t, pam_strerror(handle, error));
```

Comment:
btw, you could use `%*c` and then pass the number of chars to output, so that you don't need strndupa()